### PR TITLE
Add FilterTree/FilterSpan API on remaining analysis contexts

### PR DESCRIPTION
**Pull request by mavasani**
Work towards #68033

- Add FilterTree/FilterSpan API on all the remaing analysis contexts
- Move many IDE code style analyzers to respect the FilterTree/FilterSpan API

### All analyzers
- I carried out performance measurements for executing all analyzers for line span on few large C# files in Microsoft.CodeAnalysis.CSharp.csproj, with bits built out of current Roslyn main branch and bits built from this PR branch
- I see **~10-20% speedup** in computation (screenshot below).

image

### All high + normal priority analyzers for lightbulb (excludes SymbolStart/End and SemanticModel analyzers)
- I carried similar performance measurements as above for high + normal priority analyzers for lightbulb
- I see **~20-25% speedup** in computation (screenshot below).

image

### SymbolStart/End analyzers (known expensive analyzers)
- I carried out performance measurements for executing SymbolStart/End analyzers for line span on few large C# files in Microsoft.CodeAnalysis.CSharp.csproj, with bits built out of current Roslyn main branch and bits built from this PR branch
- I see **~10% speedup** in computation (screenshot below).
- I was hoping for a bigger improvement, but digging further I identified majority of the time is coming from method body binding to generate all the compilation events/operation trees. For example, invoking `SemanticModel.GetDiagnostics()` on `LanguageParser.cs` takes ~1.8 seconds, which is about 75% of the time (~2.2 seconds) for computing analyzer semantic diagnostics for the same file.

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> I was hoping for a bigger improvement, but digging further I identified majority of the time is coming from method body binding to generate all the compilation events/operation trees.

Can you expand on that?  Are we generating events and iop trees that are not needed?

**Comment by mavasani** [comment]
> > I was hoping for a bigger improvement, but digging further I identified majority of the time is coming from method body binding to generate all the compilation events/operation trees.
> 
> Can you expand on that? Are we generating events and iop trees that are not needed?

No, this is not redundant - the core thing is the driver has to pessimistically assume that at least one of the SymbolStart analyzer would execute on the entire symbol declaration (including all method bodies within it). Current approach for driving analyzer execution is via relevant compilation events which are generated by invoking `SemanticModel.GetDiagnostics()` on all files with named type symbol's partial definition. An alternate approach for driving analysis was attempted in #67165, but that showed the execution times were just moved to later when we create all operation trees for all method bodies. So, we would always get this hit when executing SymbolStart analyzers. The performance gain in this PR is primarily coming from the analyzers themselves doing much lesser analysis.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -4,13 +4,33 @@ Microsoft.CodeAnalysis.Compilation.SupportsRuntimeCapability(Microsoft.CodeAnaly
...
+Microsoft.CodeAnalysis.Diagnostics.OperationBlockStartAnalysisContext.FilterSpan.get -> Microsoft.CodeAnalysis.Text.TextSpan?
+Microsoft.CodeAnalysis.Diagnostics.OperationBlockStartAnalysisContext.FilterTree.get -> Microsoft.CodeAnalysis.SyntaxTree!
 Microsoft.CodeAnalysis.Diagnostics.SemanticModelAnalysisContext.FilterSpan.get -> Microsoft.CodeAnalysis.Text.TextSpan?
+Microsoft.CodeAnalysis.Diagnostics.SemanticModelAnalysisContext.FilterTree.get -> Microsoft.CodeAnalysis.SyntaxTree!
+Microsoft.CodeAnalysis.Diagnostics.SymbolAnalysisContext.FilterSpan.get -> Microsoft.CodeAnalysis.Text.TextSpan?
+Microsoft.CodeAnalysis.Diagnostics.SymbolAnalysisContext.FilterTree.get -> Microsoft.CodeAnalysis.SyntaxTree?
+Microsoft.CodeAnalysis.Diagnostics.SymbolStartAnalysisContext.FilterSpan.get -> Microsoft.CodeAnalysis.Text.TextSpan?
+Microsoft.CodeAnalysis.Diagnostics.SymbolStartAnalysisContext.FilterTree.get -> Microsoft.CodeAnalysis.SyntaxTree?
+Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext.FilterSpan.get -> Microsoft.CodeAnalysis.Text.TextSpan?
+Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext.FilterTree.get -> Microsoft.CodeAnalysis.SyntaxTree!
```
Comment (333fred) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:33]: Just warning you that you're possibly going to have merge conflicts when the public api ship for 17.6 gets here.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs [hunk 2]
```diff
@@ -25,6 +25,16 @@ internal class AnalysisScope
         public SourceOrAdditionalFile? FilterFileOpt { get; }
         public TextSpan? FilterSpanOpt { get; }
 
+        // Original filter file and filter span for the input analysis scope.
```
Comment (333fred) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs:28]: Can these be doc comments?

**Review by 333fred — APPROVED** [review]
Compiler changes seem fine to me. Have not reviewed outside the compiler layer.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs (hunk 2 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs:28]: Sure.

File: src/Analyzers/CSharp/Analyzers/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -181,6 +181,16 @@ private void AnalyzeSyntaxNode(SyntaxNodeAnalysisContext context, INamedTypeSymb
...
 
+            // If the local variable only has one declarator, then report the suggestion on the whole
+            // declaration.  Otherwise, return the suggestion only on the single declarator.
+            var reportNode = localDeclaration.Variables.Count == 1
+                ? (SyntaxNode)localDeclaration
+                : localDeclarator;
+
+            // Bail out early if the reportNode is outside the context's analysis span.
+            if (!context.ShouldAnalyzeSpan(reportNode.Span))
+                return;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs:192]: honestly, this seems like trying to be too precise.   seems like it woudl be more than fine to just do `if (!context.ShouldAnalyzeSpan(localDeclaration.Span))`.  It seems like only in rare cases would we just do a little extra work.  just my 2c.

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 4]
```diff
@@ -47,19 +57,38 @@ protected override void InitializeWorker(AnalysisContext context)
...
+                    Contract.ThrowIfNull(context.FilterTree);
+                    var shouldAnalyze = false;
+                    foreach (var member in structType.GetMembers())
+                    {
+                        var location = GetDiagnosticLocation(member, cancellationToken, out _);
+                        if (location != null && context.ShouldAnalyzeLocation(location))
+                        {
+                            shouldAnalyze = true;
+                            break;
+                        }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:83]:
```suggestion
                        if (location != null && context.ShouldAnalyzeLocation(location))
                            return true;
```

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 5]
```diff
@@ -47,19 +57,38 @@ protected override void InitializeWorker(AnalysisContext context)
...
+                    return false;
 
-                context.RegisterOperationBlockAction(
-                    context => AnalyzeBlock(context, option.Notification.Severity));
-            }, SymbolKind.NamedType);
+                // Skip analysis if the analysis filter span does not contain the primary location where we would report a diagnostic.
+                if (context.FilterSpan.HasValue)
+                {
+                    Contract.ThrowIfNull(context.FilterTree);
+                    var shouldAnalyze = false;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:75]:
```suggestion
```

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 6]
```diff
@@ -47,19 +57,38 @@ protected override void InitializeWorker(AnalysisContext context)
...
+                        var location = GetDiagnosticLocation(member, cancellationToken, out _);
+                        if (location != null && context.ShouldAnalyzeLocation(location))
+                        {
+                            shouldAnalyze = true;
+                            break;
+                        }
+                    }
+
+                    if (!shouldAnalyze)
+                        return false;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:87]:
```suggestion
                    return false;
```

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 7]
```diff
@@ -47,19 +57,38 @@ protected override void InitializeWorker(AnalysisContext context)
...
+                option = options.PreferReadOnlyStructMember;
                 if (!option.Value)
-                    return;
+                    return false;
 
-                context.RegisterOperationBlockAction(
-                    context => AnalyzeBlock(context, option.Notification.Severity));
-            }, SymbolKind.NamedType);
+                // Skip analysis if the analysis filter span does not contain the primary location where we would report a diagnostic.
+                if (context.FilterSpan.HasValue)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:72]:
```suggestion
                if (context.FilterSpan is null)
                    return true;
```

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 8]
```diff
@@ -68,18 +97,9 @@ protected override void InitializeWorker(AnalysisContext context)
...
-            {
-                MethodKind: MethodKind.Ordinary or MethodKind.ExplicitInterfaceImplementation or MethodKind.PropertyGet or MethodKind.PropertySet,
-                IsReadOnly: false,
-                IsStatic: false,
-                IsImplicitlyDeclared: false,
-                DeclaringSyntaxReferences: [var methodReference, ..],
-            } owningMethod)
-        {
+        var location = GetDiagnosticLocation(context.OwningSymbol, cancellationToken, out var additionalLocation);
+        if (location == null || !context.ShouldAnalyzeSpan(location.SourceSpan))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:101]: not understanding the `location/additionalLocation` split.

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 9]
```diff
@@ -68,18 +97,9 @@ protected override void InitializeWorker(AnalysisContext context)
...
-        if (context.OwningSymbol is not IMethodSymbol
-            {
-                MethodKind: MethodKind.Ordinary or MethodKind.ExplicitInterfaceImplementation or MethodKind.PropertyGet or MethodKind.PropertySet,
-                IsReadOnly: false,
-                IsStatic: false,
-                IsImplicitlyDeclared: false,
-                DeclaringSyntaxReferences: [var methodReference, ..],
-            } owningMethod)
-        {
+        var location = GetDiagnosticLocation(context.OwningSymbol, cancellationToken, out var additionalLocation);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:100]: could have the `IMethodSymbol` provided as an out-var here as well.

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 10]
```diff
@@ -89,10 +109,38 @@ protected override void InitializeWorker(AnalysisContext context)
             if (blockOperation is IBlockOperation { Operations: [IThrowOperation or IExpressionStatementOperation { Operation: IThrowOperation }] })
                 return;
 
-            if (BlockOperationPotentiallyMutatesThis(owningMethod, blockOperation, cancellationToken))
+            if (BlockOperationPotentiallyMutatesThis((IMethodSymbol)context.OwningSymbol, blockOperation, cancellationToken))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:112]: i'd pull this up, and do the IMethod check.  then do the other check on that imethod.

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs [hunk 11]
```diff
@@ -89,10 +109,38 @@ protected override void InitializeWorker(AnalysisContext context)
...
+            location,
+            severity,
+            additionalLocations: ImmutableArray.Create(additionalLocation),
+            properties: null));
+    }
+
+    private static Location? GetDiagnosticLocation(
+        ISymbol symbol,
+        CancellationToken cancellationToken,
+        [NotNullWhen(true)] out Location? additionalLocation)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:127]: consider a tuple-return.

File: src/Analyzers/CSharp/Analyzers/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 12]
```diff
@@ -156,17 +163,34 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext, INamedTyp
...
+                SyntaxNodeAnalysisContext context,
+                LocalDeclarationStatementSyntax localDeclaration,
+                StatementSyntax? anonymousFunctionStatement,
+                bool shouldReportOnAnonymousFunctionStatement)
+            {
+                if (context.ShouldAnalyzeSpan(localDeclaration.Span))
+                    return true;
+
+                if (shouldReportOnAnonymousFunctionStatement
+                    && context.ShouldAnalyzeSpan(anonymousFunctionStatement!.Span))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs:187]: are we trying to be far too precise here?  is there some reasonable 'top node' that we can just do all the checks on that will suffice?  like the `localDeclaration`?

File: src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpIsAndCastCheckDiagnosticAnalyzer.cs [hunk 13]
```diff
@@ -80,6 +80,11 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
+            // Bail out if the potential diagnostic location is outside the analysis span.
+            var diagnosticLocation = localDeclarationStatement.GetLocation();
+            if (!syntaxContext.ShouldAnalyzeSpan(diagnosticLocation.SourceSpan))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpIsAndCastCheckDiagnosticAnalyzer.cs:85]: same point about passing in node.

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 14]
```diff
@@ -96,10 +75,62 @@ public static void Analyze(OperationBlockStartAnalysisContext context, SymbolSta
...
+                        // issues before they've even gotten around to writing their code.
+                        if (firstBlock.IsSingleThrowNotImplementedOperation())
+                            return false;
+
+                        // If we are analyzing a specific filter tree, skip operation blocks in unrelated trees.
+                        if (symbolStartAnalyzer._symbolStartAnalysisContext.FilterTree is { } filterTree &&
+                            firstBlock.Syntax.SyntaxTree != filterTree)
+                        {
+                            return false;
+                        }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:117]: nice.

File: src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs [hunk 15]
```diff
@@ -140,6 +129,45 @@ protected sealed override void InitializeWorker(AnalysisContext context)
...
+                        var preferAutoProps = context.Options.GetAnalyzerOptions(tree).PreferAutoProperties;
+                        return preferAutoProps.Value && preferAutoProps.Notification.Severity != ReportDiagnostic.Suppress;
+                    }))
+                    {
+                        return false;
+                    }
+
+                    // If we are analyzing a sub-span (lightbulb case), then check if the filter span
+                    // has a field/property declaration where a diagnostic could be reported.
+                    if (context.FilterSpan.HasValue)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs:151]:
```suggestion
                    if (context.FilterSpan is null)
                        return true;
```

File: src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs [hunk 16]
```diff
@@ -140,6 +129,45 @@ protected sealed override void InitializeWorker(AnalysisContext context)
...
+                        Contract.ThrowIfNull(context.FilterTree);
+                        var shouldAnalyze = false;
+                        var analysisRoot = context.GetAnalysisRoot(findInTrivia: false);
+                        foreach (var node in analysisRoot.DescendantNodes())
+                        {
+                            if (node is TPropertyDeclaration or TFieldDeclaration && context.ShouldAnalyzeSpan(node.Span, node.SyntaxTree))
+                            {
+                                shouldAnalyze = true;
+                                break;
+                            }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs:162]:
```suggestion
                            if (node is TPropertyDeclaration or TFieldDeclaration && context.ShouldAnalyzeSpan(node.Span, node.SyntaxTree))
                                return true;
```

File: src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs [hunk 17]
```diff
@@ -140,6 +129,45 @@ protected sealed override void InitializeWorker(AnalysisContext context)
...
+                        }
+
+                        if (!shouldAnalyze && analysisRoot.FirstAncestorOrSelf<SyntaxNode>(node => node is TPropertyDeclaration or TFieldDeclaration) == null)
+                            return false;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs:166]:
```suggestion
                        return analysisRoot.FirstAncestorOrSelf<SyntaxNode>(node => node is TPropertyDeclaration or TFieldDeclaration) is not null;
```

File: src/Analyzers/VisualBasic/Analyzers/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedParametersAndValuesDiagnosticAnalyzer.vb [hunk 18]
```diff
@@ -21,6 +22,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.RemoveUnusedParametersAndValues
                        unusedValueAssignmentOption:=VisualBasicCodeStyleOptions.UnusedValueAssignment)
         End Sub
 
+        Protected Overrides ReadOnly Property SyntaxFacts As ISyntaxFacts
+            Get
+                Return VisualBasicSyntaxFacts.Instance
+            End Get
+        End Property
```
Comment (CyrusNajmabadi) [inline src/Analyzers/VisualBasic/Analyzers/RemoveUnusedParametersAndValues/VisualBasicRemoveUnusedParametersAndValuesDiagnosticAnalyzer.vb:29]:
```suggestion
        Protected Overrides ReadOnly Property SyntaxFacts As ISyntaxFacts = VisualBasicSyntaxFacts.Instance
```

File: src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpSimplifyTypeNamesDiagnosticAnalyzer.cs [hunk 19]
```diff
@@ -47,14 +49,16 @@ SyntaxKind.DelegateDeclaration or
...
         {
+            Debug.Assert(context.CodeBlock.DescendantNodesAndSelf().Contains(root));
+
             var semanticModel = context.SemanticModel;
             var cancellationToken = context.CancellationToken;
 
             var options = context.GetCSharpAnalyzerOptions().GetSimplifierOptions();
             using var simplifier = new TypeSyntaxSimplifierWalker(this, semanticModel, options, ignoredSpans: null, cancellationToken);
-            simplifier.Visit(context.CodeBlock);
+            simplifier.Visit(root);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpSimplifyTypeNamesDiagnosticAnalyzer.cs:61]: can we get this anlayzer to not analyze blocks unrelated to the filter span?  (if this was already done in this PR, great).

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/AnalysisContextExtensions.cs [hunk 20]
```diff
@@ -5,30 +5,144 @@
...
+
+        /// <summary>
+        /// Returns true if the given <paramref name="span"/> should be analyzed for the given <paramref name="context"/>,
+        /// i.e. either of the following is true:
+        ///  - <see cref="SymbolAnalysisContext.FilterSpan"/> is <see langword="null"/> (we are analyzing the entire tree)
+        ///  OR
+        ///  - <paramref name="span"/> intersects with <see cref="SymbolAnalysisContext.FilterSpan"/>.
+        /// </summary>
+        public static bool ShouldAnalyzeSpan(this SymbolAnalysisContext context, TextSpan span, SyntaxTree tree)
+            => context.FilterTree == null || context.FilterTree == tree && ShouldAnalyze(context.FilterSpan, span);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/AnalysisContextExtensions.cs:65]: parenthesize

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/SyntaxNodeExtensions.cs [hunk 21]
```diff
@@ -794,6 +794,17 @@ public static SyntaxToken WithoutTrailingTrivia(this SyntaxToken token)
             return token.WithTrailingTrivia(default(SyntaxTriviaList));
         }
 
+        /// <summary>
+        /// Gets the node within the given <paramref name="root"/> corresponding to the given <paramref name="span"/>.
+        /// If the <paramref name="span"/> is <see langword="null"/>, then returns the given <paramref name="root"/> node.
+        /// </summary>
+        public static SyntaxNode GetNodeForSpan(this SyntaxNode root, TextSpan? span, bool findInTrivia, bool getInnermostNodeForTie)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/SyntaxNodeExtensions.cs:801]: i feel like this is just an overload of `FindNode`.  Can we call it that, just with an optional span argument?

**Review by CyrusNajmabadi — APPROVED** [review]
signing off IDE side.  didn't look at compiler side.

slightly nervous about Location instances, but we can see in traces if its an issue.  likely we get so much savings that this is worth it.

**Review by CyrusNajmabadi — APPROVED** [review]
signing off IDE side.  didn't look at compiler side.

slightly nervous about Location instances, but we can see in traces if its an issue.  likely we get so much savings that this is worth it.

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs (hunk 4 above)
Comment (mavasani) [reply src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:83]: Same comment as previous - current code style allows us to add more bailout checks in the future without having to restructure this code.

File: src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs (hunk 8 above)
Comment (mavasani) [reply src/Analyzers/CSharp/Analyzers/MakeStructMemberReadOnly/CSharpMakeStructMemberReadOnlyAnalyzer.cs:101]: Retaining the original semantics with just refactored code. We only check whether or not the primary location is within the analysis span.

File: src/Analyzers/CSharp/Analyzers/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs (hunk 12 above)
Comment (mavasani) [reply src/Analyzers/CSharp/Analyzers/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs:187]: I am preserving the original semantics of checking the exact node on which we can report a diagnostic.

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 22]
```diff
@@ -96,10 +75,62 @@ public static void Analyze(OperationBlockStartAnalysisContext context, SymbolSta
...
+
+                        // If we are analyzing a specific filter span, skip operation blocks outside the filter span.
+                        if (context.FilterSpan.HasValue)
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:120]: Similar to my reply in one of your prior suggestions, this code style allows us to all additional bail out cases after this if statement easily, so I'll retain this pattern.

File: src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs (hunk 15 above)
Comment (mavasani) [reply src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs:151]: Same as earlier

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for a second review

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for a second review. This PR gives us an observable performance gain for lightbulb scenarios.

**Review by cston — APPROVED** [review]
Compiler changes LGTM.

**Comment by mavasani** [comment]
Thanks!

**Pull request merged**